compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -242,8 +248,8 @@ object Types {
           // If the type is `T | Null` or `T | Nothing`, and `T` derivesFrom the class,
           // then the OrType derivesFrom the class. Otherwise, we need to check both sides
           // derivesFrom the class.
-          if defn.isBottomType(tp.tp1) then loop(tp.tp2)
-          else loop(tp.tp1) && (defn.isBottomType(tp.tp2) || loop(tp.tp2))
+          loop(tp.tp1) && (loop(tp.tp2) || defn.isBottomType(tp.tp2))
+          || defn.isBottomType(tp.tp1) && loop(tp.tp2)
```

Comment:
I think the behaviour of this is same as before, but if `tp = Null | (Null | (Null | ( ... | A)`, wouldn't `tp.derivesFrom(NullClass)` cause the expensive `loop(tp.tp2)` to run two times? So I think use `if` is better here.

Reply:
Also, do we need to make sure `OrType(NullType, NothingType).derivesFrom(NothingClass) == false`?